tests/src/Clash/Vivado.hs

```diff
@@ -0,0 +1,127 @@
+{-# LANGUAGE QuasiQuotes #-}
+
+-- | Generate a TCL script to simulate generated VHDL
+--
+-- Run with @vivado -mode batch -source ...@
+module Clash.Vivado ( HdlSource (..), fromModuleName ) where
+
+import Clash.Driver.Manifest (readManifest, Manifest (..))
+import Clash.Annotations.Primitive (HDL (..))
+
+import Data.String.Interpolate (i)
+import Data.List (isSuffixOf)
+
+import System.FilePath ((</>))
+import qualified Data.Text as T
+
+data HdlSource = HdlSource { directory :: FilePath, buildTarget :: HDL }
+
+data SimProj = SimProj { vhdlFiles :: [(T.Text, FilePath)]
+                       , verilogFiles :: [(T.Text, FilePath)]
+                       , ipGen :: [FilePath]
+                       } deriving Show
+
+instance Semigroup SimProj where
+  (<>) (SimProj vhd0 v0 ip0) (SimProj vhd1 v1 ip1) =
+    SimProj (vhd0<>vhd1) (v0<>v1) (ip0<>ip1)
+
+instance Monoid SimProj where
+  mempty = SimProj [] [] []
+
+thread :: [a -> a] -> a -> a
+thread = foldr (.) id
+
+mSimProj :: FilePath -> T.Text -> FilePath -> SimProj -> SimProj
+mSimProj dir lib fp p@(SimProj vhd v ip)
+  | ".vhdl" `isSuffixOf` fp = SimProj ((lib, dir </> fp):vhd) v ip
+  | ".vhd" `isSuffixOf` fp = SimProj ((lib, dir </> fp):vhd) v ip
+  | ".v" `isSuffixOf` fp = SimProj vhd ((lib, dir </> fp):v) ip
+  | ".tcl" `isSuffixOf` fp = SimProj vhd v ((dir </> fp):ip)
+  | otherwise = p
+
+simProjFromClashEntities ::
+  -- | HDL source dir
+  FilePath ->
+  -- | Qualified name
+  String ->
+  IO SimProj
+simProjFromClashEntities hdlDir qualName = do
+  (Just mHdl) <- readManifest (hdlDir </> qualName ++ "/clash-manifest.json")
+  let deps = T.unpack <$> transitiveDependencies mHdl
+  nextSimProj <- traverse (simProjFromClashEntities hdlDir) deps
+  pure $ mconcat nextSimProj <> asSimProj qualName mHdl
+
+fromModuleName ::
+  HdlSource ->
+  -- | Module name
+  String ->
+  -- | Entity
+  String ->
+  -- | TCL script
+  IO String
+fromModuleName src moduleName entity = do
+  proj <- simProjFromClashEntities (directory src) (moduleName ++ "." ++ entity)
+  pure $ mkTcl entity src proj
+
+asSimProj :: FilePath -> Manifest -> SimProj
+asSimProj dir m =
+  let fps = fst <$> fileNames m
+    in thread (mSimProj dir (topComponent m) <$> fps) mempty
+
+mkTcl ::
+  -- | Entity
+  String ->
+  HdlSource ->
+  SimProj ->
+  String
+mkTcl entity hdl (SimProj vhdls vs tcls) =
+  [i|
+set hdl_dir #{hdlDir}
+
+# for cleanup
+file delete -force xilinx_tmp
+file mkdir xilinx_tmp
+cd xilinx_tmp
+
+create_project -in_memory
+|]
+
+  ++ concatMap (\tcl ->
+      [i|source ${hdl_dir}/#{tcl}\n|]) tcls
+
+  ++
+      [i|
+
+add_files .gen/sources_1/ip/dcfifo/synth
+
+generate_target {synthesis simulation} [get_ips {dcfifo}]
+|]
+
+  ++ readSources
+
+  ++ [i|
+synth_design -top #{entity}
+
+update_compile_order -fileset [current_fileset]
+
+set_property top #{entity} [get_fileset sim_1]
+
+save_project_as sim -force
+
+launch_simulation -simset sim_1 -mode behavioral
+
+run -all
+
+quit
```

Comment:
This is currently hardcoded for this test case. Use @DigitalBrains1's scripts?

Reply:
I have fixed it so that it doesn't need to hardcode anything but it is not as hygienic as Peter's scripts: https://github.com/clash-lang/clash-compiler/pull/2182#discussion_r899152751